hazelcast/src/main/java/com/hazelcast/spi/SplitBrainMergePolicy.java

```diff
@@ -16,14 +16,35 @@
 
 package com.hazelcast.spi;
 
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.HazelcastInstanceAware;
+import com.hazelcast.instance.Node;
+import com.hazelcast.nio.serialization.DataSerializable;
 import com.hazelcast.spi.serialization.SerializationService;
 
 /**
  * Policy for merging data structure entries after a split-brain has been healed.
+ * <p>
+ * The values of merging and existing {@link SplitBrainMergeEntryView}s are always in the in-memory format of the
+ * backing data structure. This can be a serialized format, so the content cannot be processed without deserialization.
+ * For most merge policies this will be fine, since the key or value are not used.
+ * A merge policy can implement {@link NodeAware} to get an instance of {@link Node} injected.
+ * Via {@link Node#getSerializationService()} the {@link SerializationService} can be retrieved to deserialize the data.
+ * <p>
+ * The deserialization is not done eagerly for two main reasons:
```

Comment:
can we preserve this option(do not deserialize eagerly) and change default to always deserialize? Having deserialized seems the default expectation to me. Merge policies in previous versions are like this.

Reply:
The `IMap` and `ICache` had lazy deserializiation per default. Only the `ReplicatedMap` deserialized eagerly, which is a _huge_ waste of resources. No out-of-the-box merge policy uses the key or value, so a default of always deserialize makes no sense to me.

It's so easy to get your hands on the `SerializationSerivce` now, that I don't think it's a big thing to implement this in your custom merge policy:
```java
    protected static class MergeIntegerValuesMergePolicy implements SplitBrainMergePolicy, NodeAware {

        private transient SerializationService serializationService;

        @Override
        public <T> T merge(MergeDataHolder<T> mergingData, MergeDataHolder<T> existingData) {
            if (serializationService.toObject(mergingData.getValue()) instanceof Integer) {
                return mergingData.getValue();
            }
            return null;
        }

        @Override
        public void setNode(Node node) {
            this.serializationService = node.getSerializationService();
        }

        @Override
        public void writeData(ObjectDataOutput out) {
        }

        @Override
        public void readData(ObjectDataInput in) {
        }
    }
```

Reply:
I was meaning requirement for SerializationService. But also lazy deserialization in current form can be problematic for hd data, if one offloads some related processing to a new thread inside merge method and tries to access data later --> JVM can crash, seems by default only on heap data should be given to user